providers/keycloak.go

```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
 	p.Group = group
 }
 
-func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	json, err := requests.New(p.ValidateURL.String()).
+func (p *KeycloakProvider) SetRoles(roles []string) {
+	p.Roles = roles
+}
+
+type keycloakUserInfo struct {
+	Email  string   `json:"email"`
+	Groups []string `json:"groups"`
+}
+
+func (p *KeycloakProvider) getUserInfo(ctx context.Context, s *sessions.SessionState) (*keycloakUserInfo, error) {
+	var userInfo keycloakUserInfo
+	err := requests.New(p.ValidateURL.String()).
 		WithContext(ctx).
 		SetHeader("Authorization", "Bearer "+s.AccessToken).
 		Do().
-		UnmarshalJSON()
+		UnmarshalInto(&userInfo)
 	if err != nil {
 		logger.Errorf("failed making request %s", err)
-		return "", err
+		return nil, fmt.Errorf("error getting user info: %v", err)
+	}
+	return &userInfo, nil
+}
+
+func getClientRoles(resourceAccess map[string]interface{}) ([]string, error) {
+	var clientRoles []string
+	for name, list := range resourceAccess {
+		scopes, ok := list.(map[string]interface{})
+		if !ok {
+			return nil, errors.New("error parsing client roles from claims")
+		}
+		if roles, found := scopes["roles"]; found {
+			for _, r := range roles.([]interface{}) {
+				clientRoles = append(clientRoles, fmt.Sprintf("%s:%s", name, r))
+			}
+		}
 	}
+	return clientRoles, nil
+}
+
+type realmAccess struct {
+	Roles []string `json:"roles"`
+}
+
+type customClaims struct {
+	RealmAccess    realmAccess            `json:"realm_access"`
+	ResourceAccess map[string]interface{} `json:"resource_access"`
+}
+
+func (p *KeycloakProvider) getRoles(s *sessions.SessionState) ([]string, error) {
+	// Decode JWT token without verifying the signature
+	token, err := jwt.ParseSigned(s.AccessToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse token: %v", err)
+	}
+	standardClaims := jwt.Claims{}
+	customClaims := customClaims{}
+	// Parse claims
+	if err := token.UnsafeClaimsWithoutVerification(&standardClaims, &customClaims); err != nil {
+		logger.Printf("failed to parse claims: %s", err)
+	}
+	clientRoles, err := getClientRoles(customClaims.ResourceAccess)
+	if err != nil {
+		logger.Printf("failed to extract client roles: %s", err)
+	}
+	return append(customClaims.RealmAccess.Roles, clientRoles...), nil
```

Comment:
It feels weird to pollute & extend the underlying object we extracted from the token claims.

I understand the performance motivations of reusing the existing slice since it is done at the end of the function anyway, but since these are small & therefore performance is negligible, maybe clarity is more ideal? Took me a second on first glance to realize the desired output was intended as `RealmAccess` + ResourceAccess`

Reply:
You're right, I will change this for better clarity.